src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,458 @@
+            }
+        }
+    }
+
+    /// Send a request to the server.
+    pub async fn send_data(&self, data: &[u8]) -> Result<(), RpcError<String>> {
+        let writer = self.get_stream_mut();
+        match write_all_timeout!(writer, data, self.timeout_options.write_timeout).await {
+            Ok(_) => {
+                debug!("Sent data with length: {:?}", data.len());
+                return Ok(());
+            }
+            Err(err) => {
+                debug!("Failed to send data: {:?}", err);
+                return Err(RpcError::InternalError(err.to_string()));
+            }
+        }
+    }
+
+    /// Get the next sequence number.
+    pub fn next_seq(&self) -> u64 {
+        self.seq.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+    }
+
+    pub async fn ping(&self) -> Result<(), RpcError<String>> {
+        // Send keep alive message
+        let current_seq = self.next_seq();
+        // Check keepalive is valid
+        let received_keepalive_seq = self.received_keepalive_seq.load(std::sync::atomic::Ordering::Relaxed);
+        if current_seq - received_keepalive_seq > 100 {
+            debug!("Keep alive timeout, close the connection");
+            return Err(RpcError::InternalError("Keep alive timeout".to_string()));
+        }
+
+        // Set to packet task
+        let keep_alive_msg = ReqHeader {
+            seq: current_seq,
+            op: ReqType::KeepAliveRequest.to_u8(),
+            len: 0,
+        }.encode();
+
+        if let Ok(_) =  self.send_data(&keep_alive_msg).await {
+            debug!("Success to sent keep alive message");
+            Ok(())
+        } else {
+            debug!("Failed to send keep alive message");
+            Err(RpcError::InternalError("Failed to send keep alive message".to_string()))
+        }
+    }
+
+    /// Send packet by the client.
+    pub async fn send_packet(&self, req_packet: &mut P) -> Result<(), RpcError<String>> {
+        let current_seq = self.next_seq();
+        req_packet.set_seq(current_seq);
+        debug!("Try to send request: {:?}", current_seq);
+
+        // Send may be used in different threads, so we need to create local buffer to store the request data
+        if let Ok(req_buffer) = req_packet.get_req_data() {
+            let mut req_header = ReqHeader {
+                seq: req_packet.seq(),
+                op: req_packet.op(),
+                len: req_buffer.len() as u64,
+            }.encode();
+
+            // concate req_header and req_buffer
+            req_header.extend_from_slice(&req_buffer);
+
+            if let Ok(_) = self.send_data(&req_header).await {
+                debug!("Sent request success: {:?}", req_packet.seq());
+                // We have set a copy to keeper and manage the status for the packets keeper
+                // Set to packet task with clone
+                self.get_packets_keeper_mut().add_task(req_packet.clone()).await;
```

Comment:
add_task() does not need to be async fn.

Reply:
Ok, fixed.